[AMDGPU] Extend readlane, writelane and readfirstlane intrinsic lowering for generic types

vikramRH opened the pull request:
This patch is intended to be the first of a series with end goal to adapt atomic optimizer pass to support i64 and f64 operations (along with removing all unnecessary bitcasts). This patch introduces pseudos for 64bit readlane, writelane and readfirstlane ops which are lowered post ISel. Requesting comments on the approach here,

vikramRH committed:
[AMDGPU] add support for i64 readlane

vikramRH committed:
add support for i64 readfirstlane and writelane intrinsics

vikramRH committed:
Fix issues with writelane expansion

vikramRH committed:
code refactor and add patterns for f64

vikramRH committed:
clang format

vikramRH committed:
fix corner case with regkill and add readlane tests

vikramRH committed:
update builtin handling for readlane and readfirstlane

vikramRH committed:
add and update tests, fixes to writelane src0 imm handling

vikramRH commented:
Added/updated tests for readfirstlane and writelane ops

vikramRH commented:
Gentle ping :)

arsenm commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -18410,6 +18410,24 @@ Value *CodeGenFunction::EmitAMDGPUBuiltinExpr(unsigned BuiltinID,
+    getContext().GetBuiltinType(BuiltinID, Error, &ICEArguments);
+    assert(Error == ASTContext::GE_None && "Should not codegen an error");
+    for (unsigned I = 0; I != E->getNumArgs(); ++I) {
+      Args.push_back(EmitScalarOrConstFoldImmArg(ICEArguments, I, E));
```
This doesn't require the constant folding. This can use a plain EmitScalarExpr

arsenm commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -18410,6 +18410,24 @@ Value *CodeGenFunction::EmitAMDGPUBuiltinExpr(unsigned BuiltinID,
     return Builder.CreateCall(F, Args);
   }
+  case AMDGPU::BI__builtin_amdgcn_readlane:
+  case AMDGPU::BI__builtin_amdgcn_readfirstlane: {
```
readfirstlane case can just call emitUnaryBuiltin. readlane can probably get away with emitBinaryBuiltin, although it doesn't quite fit the binary intrinsic signature

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -4822,6 +4822,111 @@ static MachineBasicBlock *lowerWaveReduce(MachineInstr &MI,
   return RetBB;
 }
 
+static MachineBasicBlock *lowerPseudoLaneOp(MachineInstr &MI,
```
You should try to do this before selection. Doing it after just adds a lot of complexity and interferes with combine opportunities

vikramRH replied:
@arsenm, would "PreISelIntrinsicLowering" be a proper place for this ?

arsenm replied:
No, that's a generic pass. I would directly handle this in the legalizer, in SIISelLowering and AMDGPULegalizer

arsenm reviewed (commented):
In a separate patch should have AMDGPUInstCombineIntrinsic try to fold bitcasts into the intrinsic

Commit:
address review comments

jayfoad commented:
Previous attempts:
* https://reviews.llvm.org/D84639
* https://reviews.llvm.org/D86154
* https://reviews.llvm.org/D147732
* #87334

vikramRH committed:
Implement lowering in legalizer for legal types

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUInstrInfo.td:
```diff
@@ -504,3 +508,15 @@ def AMDGPUdiv_fmas : PatFrags<(ops node:$src0, node:$src1, node:$src2, node:$vcc
+
+def AMDGPUwritelane : PatFrags<(ops node:$src0, node:$src1, node:$src2),
+  [(int_amdgcn_writelane node:$src0, node:$src1, node:$src2),
+   (AMDGPUwritelane_impl node:$src0, node:$src1, node:$src2)]>;
```
Missing newline end of file

vikramRH commented:
new commit extends @jayfoad's implementation with GIsel support.  yet to add tests for half, floats and some vectors

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -5386,6 +5386,94 @@ bool AMDGPULegalizerInfo::legalizeDSAtomicFPIntrinsic(LegalizerHelper &Helper,
+    }
+    B.buildTrunc(DstReg, LaneOpDst).getReg(0);
+  } else if ((Size % 32) == 0) {
+    SmallVector<Register, 2> Src0Parts, PartialRes;
+    unsigned NumParts = Size / 32;
+    auto WideReg = MRI.createGenericVirtualRegister(LLT::scalar(NumParts * 32));
+    for (unsigned i = 0; i < NumParts; ++i) {
+      Src0Parts.push_back(MRI.createGenericVirtualRegister(S32));
+    }
+
+    B.buildUnmerge(Src0Parts, Src0);
```
buildUnmerge should handle all of this for you if you just pass the scalar type

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -5386,6 +5386,94 @@ bool AMDGPULegalizerInfo::legalizeDSAtomicFPIntrinsic(LegalizerHelper &Helper,
+        LaneOpDst = LaneOpDst.addUse(Ext2);
+      }
+    }
+    B.buildTrunc(DstReg, LaneOpDst).getReg(0);
```
Should just early exit at this point

arsenm replied:
The .getReg(0) does nothing here

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -6091,6 +5982,70 @@ static SDValue lowerBALLOTIntrinsic(const SITargetLowering &TLI, SDNode *N,
+
+  auto createLaneOp = [&](SDValue &Src0, SDValue &Src1, SDValue &Src2,
+                          MVT VT) -> SDValue {
+    return (Src2.getNode()
```
Don't need .getNode for boolean test

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -6091,6 +5982,70 @@ static SDValue lowerBALLOTIntrinsic(const SITargetLowering &TLI, SDNode *N,
+      Src2Valid = DAG.getBitcast(VecVT, Src2);
+
+    auto LaneOp = createLaneOp(Src0Valid, Src1, Src2Valid, VecVT);
+    auto UnrolledLaneOp = DAG.UnrollVectorOp(LaneOp.getNode());
```
no autos

vikramRH committed:
refactor/improve GIsel lowering, added new tests

1. Review comments
2. improve GIsel lowering
3. add tests for half, bfloat, float2, ptr, vector of ptr and int
4. removed gfx700 checks from writelane tests since it caused issues with f16 legalization. is this required ?

vikramRH commented:
1. Review comments
2. improve GIsel lowering
3. add tests for half, bfloat, float2, ptr, vector of ptr and int
4. removed gfx700 checks from writelane test since it caused issues with f16 legalization. is this required ?

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -5982,6 +5982,68 @@ static SDValue lowerBALLOTIntrinsic(const SITargetLowering &TLI, SDNode *N,
+  SDLoc SL(N);
+  MVT IntVT = MVT::getIntegerVT(ValSize);
+
+  auto createLaneOp = [&](SDValue &Src0, SDValue &Src1, SDValue &Src2,
```
SDValue should be passed by value

vikramRH committed:
Review comments, refactor GISel Impl

vikramRH committed:
clang-format

vikramRH commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -5386,6 +5386,130 @@ bool AMDGPULegalizerInfo::legalizeDSAtomicFPIntrinsic(LegalizerHelper &Helper,
+      Register Src1 = MI.getOperand(3).getReg();
+      for (unsigned i = 0; i < NumParts; ++i)
+        PartialRes.push_back(
+            (B.buildIntrinsic(Intrinsic::amdgcn_readlane, {S32})
+                 .addUse(Src0Parts.getReg(i))
+                 .addUse(Src1))
+                .getReg(0));
```
should this be a seperate change that addresses other such instances too ?

jayfoad replied:
Yes, separate patch

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -5386,6 +5386,153 @@ bool AMDGPULegalizerInfo::legalizeDSAtomicFPIntrinsic(LegalizerHelper &Helper,
+
+    Register Src0Valid = B.buildBitcast(S32, Src0).getReg(0);
+    MachineInstrBuilder LaneOpDst;
+    switch (IID) {
```
Can you use a `createLaneOp` helper to build the intrinsic, like you do in the SDAG path?

vikramRH replied:
I removed the helper in the recent commit following @arsenm's suggestion. Only reason is readability

arsenm replied:
This isn't quite what I meant; I mean trying to handle these all as they were the same create lane was ugly

vikramRH replied:
my bad, I will improve the helper

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -5386,6 +5386,153 @@ bool AMDGPULegalizerInfo::legalizeDSAtomicFPIntrinsic(LegalizerHelper &Helper,
+      // Already legal
+      return true;
+
+    Register Src0Valid = B.buildBitcast(S32, Src0).getReg(0);
```
I think you can just reuse Src0 instead of declaring a new Src0Valid. Same for Src2, and same for the SDAG code.

jayfoad reviewed (commented):
LGTM overall.

> add f32 pattern to select read/writelane operations

Why would you need this? Don't you legalize f32 to i32?

vikramRH commented:
> > add f32 pattern to select read/writelane operations
> 
> Why would you need this? Don't you legalize f32 to i32?

Sorry about this. Its a leftover comment from the initial implementation which I should have removed.

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUAtomicOptimizer.cpp:
```diff
@@ -493,8 +493,8 @@ Value *AMDGPUAtomicOptimizerImpl::buildScan(IRBuilder<> &B,
       V = B.CreateBitCast(V, IntNTy);
-      Value *const Lane31 = B.CreateIntrinsic(Intrinsic::amdgcn_readlane, {},
-                                              {V, B.getInt32(31)});
+      Value *const Lane31 = B.CreateIntrinsic(
+          Intrinsic::amdgcn_readlane, B.getInt32Ty(), {V, B.getInt32(31)});
```
Changes like this should disappear if we merge #91583 first.

arsenm commented on llvm/include/llvm/IR/IntrinsicsAMDGPU.td:
```diff
@@ -2176,26 +2176,23 @@ def int_amdgcn_wave_reduce_umin : AMDGPUWaveReduce;
-    llvm_i32_ty,    // uniform lane select
-    llvm_i32_ty     // returned by all lanes other than the selected one
+  Intrinsic<[llvm_any_ty], [
+    LLVMMatchType<0>,    // uniform value to write: returned by the selected lane
+    llvm_i32_ty,        // uniform lane select
+    LLVMMatchType<0>     // returned by all lanes other than the selected one
```
Comments are no longer aligned

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -5386,6 +5386,153 @@ bool AMDGPULegalizerInfo::legalizeDSAtomicFPIntrinsic(LegalizerHelper &Helper,
+      LaneOpDst = B.buildIntrinsic(IID, {S32})
+                      .addUse(Src0Valid)
+                      .addUse(Src1)
+                      .addUse(Src2Valid);
```
Missing break

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -5386,6 +5386,153 @@ bool AMDGPULegalizerInfo::legalizeDSAtomicFPIntrinsic(LegalizerHelper &Helper,
+    if (Ty.isPointerVector()) {
+      auto MergedVec = B.buildMergeLikeInstr(
+          LLT::vector(ElementCount::getFixed(NumParts), S32), PartialRes);
+      B.buildBitcast(DstReg, MergedVec);
```
You cannot bitcast from an i32 vector to a pointer. You can merge the i32 pieces directly into the pointer though

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -5386,6 +5386,153 @@ bool AMDGPULegalizerInfo::legalizeDSAtomicFPIntrinsic(LegalizerHelper &Helper,
+  if ((Size % 32) == 0) {
+    SmallVector<Register, 2> PartialRes;
+    unsigned NumParts = Size / 32;
+    auto Src0Parts = B.buildUnmerge(S32, Src0);
```
For the multiple of `<2 x s16>`, it's a bit nicer to preserve the 16-bit element types

vikramRH replied:
Do you mean extract s16 elements individually and handle them as  (Size < 32) case ?

vikramRH replied:
done I hope as per the expectation, however I don't understand the plus  here

arsenm replied:
You avoid adding extra intermediate bitcasts

vikramRH committed:
Merge branch 'main' into rw_lane_64

vikramRH committed:
Review comments, improve pointer handling with GISel

vikramRH committed:
align comments

vikramRH commented:
Added new 32 bit pointer,  <8 x i16> tests

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -5387,6 +5387,212 @@ bool AMDGPULegalizerInfo::legalizeDSAtomicFPIntrinsic(LegalizerHelper &Helper,
+
+  if (Size == 32) {
+    if (Ty.isScalar())
+      // Already legal
```
Either add braces or don't put the comment under the if. Also, the size 32 case should be treated as directly legal for 32-bit pointers

vikramRH replied:
Done except for pointers. I currently see an issue where pattern type inference somehow deduces destination type to scalars (instead of say LLT_ p3s32). not currently sure why , any ideas ?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -5387,6 +5387,212 @@ bool AMDGPULegalizerInfo::legalizeDSAtomicFPIntrinsic(LegalizerHelper &Helper,
+      return true;
+
+    auto IsPtr = Ty.isPointer();
+    Src0 = IsPtr ? B.buildPtrToInt(S32, Src0).getReg(0)
+                 : B.buildBitcast(S32, Src0).getReg(0);
```
You should not need these casts. Any legal type should be directly accepted without these

arsenm commented on llvm/lib/Target/AMDGPU/SIInstructions.td:
```diff
@@ -3400,7 +3400,7 @@ def : GCNPat<
 // the ignored src1.
 def : GCNPat<
-  (int_amdgcn_readfirstlane (i32 imm:$src)),
+  (i32 (AMDGPUreadfirstlane (i32 imm:$src))),
```
We might need to make this fold more sophisticated for other types, but best for a follow up patch

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -6086,6 +6086,68 @@ static SDValue lowerBALLOTIntrinsic(const SITargetLowering &TLI, SDNode *N,
+    if (VT == MVT::i32)
+      // Already legal
+      return SDValue();
+    Src0 = DAG.getBitcast(IntVT, Src0);
```
Like the other cases, we should be able to avoid intermediate casting

vikramRH committed:
Review comments

arsenm commented on llvm/lib/Target/AMDGPU/VOP1Instructions.td:
```diff
@@ -243,11 +243,16 @@ def VOP_READFIRSTLANE : VOPProfile <[i32, i32, untyped, untyped]> {
 
+foreach vt = Reg32Types.types in {
+  def : GCNPat<(vt (AMDGPUreadfirstlane (vt VRegOrLdsSrc_32:$src0))),
+        (V_READFIRSTLANE_B32 (vt VRegOrLdsSrc_32:$src0))
```
```suggestion
        (vt (V_READFIRSTLANE_B32 (vt VRegOrLdsSrc_32:$src0)))
```

This may fix your pattern type deduction issue

vikramRH replied:
Unfortunately no, Had tried this and couple of other variations. the issue seems to be too specific to GIsel pointers..

arsenm replied:
What is the problem?

vikramRH replied:
Attaching example match table snippets for v2i16 and p3 here, should make the scenario bit more clear,
for v2i16
 ```
GIM_Try, /*On fail goto*//*Label 3499*/ GIMT_Encode4(202699), // Rule ID 2117 //
        GIM_CheckIntrinsicID, /*MI*/0, /*Op*/1, GIMT_Encode2(Intrinsic::amdgcn_writelane),
        GIM_RootCheckType, /*Op*/0, /*Type*/GILLT_v2s16,
        GIM_RootCheckType, /*Op*/2, /*Type*/GILLT_v2s16,
        GIM_RootCheckType, /*Op*/3, /*Type*/GILLT_s32,
        GIM_RootCheckType, /*Op*/4, /*Type*/GILLT_v2s16,
        GIM_RootCheckRegBankForClass, /*Op*/0, /*RC*/GIMT_Encode2(AMDGPU::VGPR_32RegClassID),
        // (intrinsic_wo_chain:{ *:[v2i16] } 2863:{ *:[iPTR] }, v2i16:{ *:[v2i16] }:$src0, i32:{ *:[i32] }:$src1, v2i16:{ *:[v2i16] }:$src2)  =>  (V_WRITELANE_B32:{ *:[v2i16] } SCSrc_b32:{ *:[v2i16] }:$src0, SCSrc_b32:{ *:[i32] }:$src1, VGPR_32:{ *:[v2i16] }:$src2)
        GIR_BuildRootMI, /*Opcode*/GIMT_Encode2(AMDGPU::V_WRITELANE_B32),
```

and for p3,
```
GIM_Try, /*On fail goto*//*Label 3502*/ GIMT_Encode4(202816), // Rule ID 2129 //
        GIM_CheckIntrinsicID, /*MI*/0, /*Op*/1, GIMT_Encode2(Intrinsic::amdgcn_writelane),
        GIM_RootCheckType, /*Op*/0, /*Type*/GILLT_s32,
        GIM_RootCheckType, /*Op*/2, /*Type*/GILLT_p2s32,
        GIM_RootCheckType, /*Op*/3, /*Type*/GILLT_s32,
        GIM_RootCheckType, /*Op*/4, /*Type*/GILLT_p2s32,
        GIM_RootCheckRegBankForClass, /*Op*/0, /*RC*/GIMT_Encode2(AMDGPU::VGPR_32RegClassID),
        // (intrinsic_wo_chain:{ *:[i32] } 2863:{ *:[iPTR] }, p2:{ *:[i32] }:$src0, i32:{ *:[i32] }:$src1, p2:{ *:[i32] }:$src2)  =>  (V_WRITELANE_B32:{ *:[i32] } SCSrc_b32:{ *:[i32] }:$src0, SCSrc_b32:{ *:[i32] }:$src1, VGPR_32:{ *:[i32] }:$src2)
        GIR_BuildRootMI, /*Opcode*/GIMT_Encode2(AMDGPU::V_WRITELANE_B32),
```

The destination type check for p3 case is still for "GILLT_s32",

arsenm replied:
Seems like this is just a GlobalISelEmitter bug (and another example of we we should have a way to write patterns that only care about the bit size)

arsenm replied:
I think GlobalISelEmitter is just ignoring PtrValueType

vikramRH replied:
Do you think these changes are okay until I figure out root cause ?

arsenm replied:
I'd rather just leave the pointer cases failing in the global isel case, and remove all the cast insertion bits.
You could split out the pointer cases to a separate file and just not run it with globalisel.

vikramRH replied:
Done

vikramRH committed:
fix type profile

vikramRH commented on llvm/lib/Target/AMDGPU/AMDGPUInstrInfo.td:
```diff
@@ -342,6 +342,22 @@ def AMDGPUfdot2_impl : SDNode<"AMDGPUISD::FDOT2",
+]>;
+
+def AMDGPUDWritelaneOp : SDTypeProfile<1, 3, [
+  SDTCisSameAs<1, 1>, SDTCisInt<2>, SDTCisSameAs<0, 3>,
```
Thanks for pointing this, missed updating this latest version. updated now, however issue is not related to this

vikramRH committed:
remove spurious comma

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -6086,6 +6086,62 @@ static SDValue lowerBALLOTIntrinsic(const SITargetLowering &TLI, SDNode *N,
+  SDLoc SL(N);
+  MVT IntVT = MVT::getIntegerVT(ValSize);
+
+  auto createLaneOp = [&](SDValue Src0, SDValue Src1, SDValue Src2,
+                          MVT VT) -> SDValue {
```
&DAG? VT shadow

vikramRH committed:
review comments, move pointer tests to new files

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -5387,6 +5387,192 @@ bool AMDGPULegalizerInfo::legalizeDSAtomicFPIntrinsic(LegalizerHelper &Helper,
+  if ((Size % 32) == 0) {
+    SmallVector<Register, 2> PartialRes;
+    unsigned NumParts = Size / 32;
+    auto IsS16Vec = Ty.isVector() && Ty.getElementType() == S16;
```
no auto

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -5387,6 +5387,192 @@ bool AMDGPULegalizerInfo::legalizeDSAtomicFPIntrinsic(LegalizerHelper &Helper,
+    } else {
+      if (IsS16Vec) {
+        for (unsigned i = 0; i < NumParts; i++)
+          PartialRes[i] = B.buildBitcast(V2S16, PartialRes[i]).getReg(0);
```
You shouldn't need to create a bitcast here. You're directly consuming the `<2 x s16>` pieces in the selection pattern

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -6086,6 +6086,62 @@ static SDValue lowerBALLOTIntrinsic(const SITargetLowering &TLI, SDNode *N,
+    SDValue InitBitCast = DAG.getBitcast(IntVT, Src0);
+    Src0 = DAG.getAnyExtOrTrunc(InitBitCast, SL, MVT::i32);
+    if (Src2.getNode()) {
+      SDValue Src2Cast = DAG.getBitcast(IntVT, Src2);
```
You should not have any bitcasts anywhere

vikramRH replied:
What would be the proper way to legalize f16 and bf16 for SDAG case without bitcasts ? (Im currently thinking  "fp_extend -> LaneOp -> Fptrunc" which seems wrong)

arsenm replied:
Yes, bitcast for the f16/bf16 case to get to the int

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -5387,6 +5387,192 @@ bool AMDGPULegalizerInfo::legalizeDSAtomicFPIntrinsic(LegalizerHelper &Helper,
+      Register Src2 = MI.getOperand(4).getReg();
+      MachineInstrBuilder Src2Parts;
+
+      if (Ty.isPointer()) {
+        auto PtrToInt = B.buildPtrToInt(S64, Src2);
+        Src2Parts = B.buildUnmerge(S32, PtrToInt);
+      } else if (Ty.isPointerVector()) {
+        LLT IntVecTy = Ty.changeElementType(
+            LLT::scalar(Ty.getElementType().getSizeInBits()));
+        auto PtrToInt = B.buildPtrToInt(IntVecTy, Src2);
+        Src2Parts = B.buildUnmerge(S32, PtrToInt);
+      } else
+        Src2Parts =
+            IsS16Vec ? B.buildUnmerge(V2S16, Src2) : B.buildUnmerge(S32, Src2);
```
The point of splitting out the pointer typed tests was to avoid fixing the handling of the pointer typed selection patterns. You still have the casts inserted here. You should not need any ptrtoint, inttoptr, or bitcasts in any of these legalizations

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -5387,6 +5387,192 @@ bool AMDGPULegalizerInfo::legalizeDSAtomicFPIntrinsic(LegalizerHelper &Helper,
+    }
+    case Intrinsic::amdgcn_readfirstlane: {
+      for (unsigned i = 0; i < NumParts; ++i) {
+        Src0 = IsS16Vec ? B.buildBitcast(S32, Src0Parts.getReg(i)).getReg(0)
```
No bitcasts

arsenm reviewed (changes requested):
There should be no need to introduce same-sized value casts, whether bitcast or ptrtoint in either legalizer

vikramRH committed:
remove bitcasts, avoid special handling of pointers in gisel

vikramRH commented:
updated the GIsel legalizer, I still have couple of questions for SDAG case though,
1. What's the proper way to legalize f16 and bf16 for SDAG case without bitcasts ? (I would think  "fp_extend -> LaneOp -> Fptrunc" is wrong)
2. For scalar cases such as i64, f64, i128 .. (i.e 32 bit multiples), I guess bitcast to vectors (v2i32, v2f32, v4i32) is unavoidable since "UnrollVectorOp" wouldn't work otherwise. any alternalte suggestions here ?

jayfoad commented:
> 1. What's the proper way to legalize f16 and bf16 for SDAG case without bitcasts ? (I would think  "fp_extend -> LaneOp -> Fptrunc" is wrong)

Bitcast to i16, anyext to i32, laneop, trunc to i16, bitcast to original type.

Why wouldn't you use bitcasts?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -5456,43 +5444,32 @@ bool AMDGPULegalizerInfo::legalizeLaneOp(LegalizerHelper &Helper,
     SmallVector<Register, 2> PartialRes;
     unsigned NumParts = Size / 32;
-    auto IsS16Vec = Ty.isVector() && Ty.getElementType() == S16;
+    bool IsS16Vec = Ty.isVector() && Ty.getElementType() == S16;
```
Better to track this as the LLT to use for the pieces, rather than making it this conditional thing. This will simplify improved pointer handling in the future

vikramRH replied:
done

vikramRH commented:
> > 1. What's the proper way to legalize f16 and bf16 for SDAG case without bitcasts ? (I would think  "fp_extend -> LaneOp -> Fptrunc" is wrong)
> 
> Bitcast to i16, anyext to i32, laneop, trunc to i16, bitcast to original type.
> 
> Why wouldn't you use bitcasts?

Just a doubt I had on previous comments, sorry for the noise !

arsenm reviewed (commented):
Should lose the [WIP] in the title

vikramRH committed:
Review comments, updated AMDGPUUsage.rst

vikramRH commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -6086,6 +6086,63 @@ static SDValue lowerBALLOTIntrinsic(const SITargetLowering &TLI, SDNode *N,
+  }
+
+  if ((ValSize % 32) == 0) {
+    MVT VecVT = MVT::getVectorVT(MVT::i32, ValSize / 32);
+    Src0 = DAG.getBitcast(VecVT, Src0);
+
+    if (Src2.getNode())
+      Src2 = DAG.getBitcast(VecVT, Src2);
+
+    SDValue LaneOp = createLaneOp(Src0, Src1, Src2, VecVT);
+    SDValue UnrolledLaneOp = DAG.UnrollVectorOp(LaneOp.getNode());
+    return DAG.getBitcast(VT, UnrolledLaneOp);
```
```suggestion
  MVT LaneOpT =
        VT.isVector() && VT.getVectorElementType().getSizeInBits() == 16
            ? MVT::v2i16
            : MVT::i32;
    SDValue Src0SubReg, Src2SubReg;
    SmallVector<SDValue, 4> LaneOps;
    LaneOps.push_back(DAG.getTargetConstant(
        TLI.getRegClassFor(VT.getSimpleVT(), N->isDivergent())->getID(), SL,
        MVT::i32));
    for (unsigned i = 0; i < (ValSize / 32); i++) {
      unsigned SubRegIdx = SIRegisterInfo::getSubRegFromChannel(i);
      Src0SubReg = DAG.getTargetExtractSubreg(SubRegIdx, SL, LaneOpT, Src0);
      if (Src2)
        Src2SubReg = DAG.getTargetExtractSubreg(SubRegIdx, SL, LaneOpT, Src2);
      LaneOps.push_back(createLaneOp(Src0SubReg, Src1, Src2SubReg, LaneOpT));
      LaneOps.push_back(DAG.getTargetConstant(SubRegIdx, SL, MVT::i32));
    }
    return SDValue(
        DAG.getMachineNode(TargetOpcode::REG_SEQUENCE, SL, VT, LaneOps), 0);
```

@arsenm , @jayfoad , an alternate idea here that is much closer in logic to the GIsel implementation and doesn't rely on bitcasts. how does this look ?

arsenm replied:
You should not be using TLI.getRegClassFor or getTargetExtractSubreg. The GlobalISel equivalent does not use these. Stick to the higher level operations unless necessary, which it isn't here

arsenm commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -1170,6 +1170,23 @@ The AMDGPU backend implements the following LLVM IR intrinsics.
 
+  llvm.amdgcn.readfirstlane                        Provides direct access to v_readfirstlane_b32. Returns the value in
+                                                   the lowest active lane of the input operand. Currently 
+                                                   implemented for i16, i32, float, half, bf16, v2i16, v2f16 and types 
```
This mixes type naming conventions. Probably should stick to the IR convention (and say `bfloat`, `<2 x i16>`, `<2 x half>`, `<2 x bfloat>`. Also, pointers double, and multiples of the 32-bit vectors should work.

arsenm commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -1170,6 +1170,23 @@ The AMDGPU backend implements the following LLVM IR intrinsics.
+  llvm.amdgcn.readlane                             Provides direct access to v_readlane_b32. Returns the value in the 
+                                                   specified lane of the first input operand. The second operand 
+                                                   specifies the lane to read from. Currently implemented
+                                                   for i16, i32, float, half, bf16, v2i16, v2f16 and types whose sizes
```
Should probably try to avoid repeating all the same information on all 3 intrinsics

vikramRH replied:
Updated

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -5387,6 +5387,124 @@ bool AMDGPULegalizerInfo::legalizeDSAtomicFPIntrinsic(LegalizerHelper &Helper,
+    return true;
+  }
+
+  if ((Size % 32) == 0) {
```
Early return instead of indenting everything on this.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -5387,6 +5387,124 @@ bool AMDGPULegalizerInfo::legalizeDSAtomicFPIntrinsic(LegalizerHelper &Helper,
+  if ((Size % 32) == 0) {
+    SmallVector<Register, 2> PartialRes;
+    unsigned NumParts = Size / 32;
+    LLT PartialResTy =
+        Ty.isVector() && Ty.getElementType() == S16 ? V2S16 : S32;
```
This isn't trying to preserve pointer element types, and will also lose FP information in the future. Can you use changeElementCount in the size == 16 case, and try to maintain the element type for the 32-bit case?

vikramRH replied:
Done

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -6086,6 +6086,63 @@ static SDValue lowerBALLOTIntrinsic(const SITargetLowering &TLI, SDNode *N,
+  }
+
+  if ((ValSize % 32) == 0) {
+    MVT VecVT = MVT::getVectorVT(MVT::i32, ValSize / 32);
```
Same comment, but more important than the globalisel case. Try to maintain the element type (e.g. v2f32->f32)

vikramRH replied:
So relying on vector bitcasts is okay ? also thinking about types such as f64, i128.. , How should these be handled ?

arsenm replied:
Yes, you need bitcast to get from FP to the 32-bit scalars. you are only trying to preserve the element types of the 32-bit legal pieces. i.e. v4f16 -> v2f16, v2f16. v4f32 -> f32, f32, f32, f32. v2f64 -> bitcast v4i32

vikramRH replied:
Understood. Thanks !

vikramRH replied:
Updated

vikramRH committed:
preserve legel 32-bit pieces, update usage doc

vikramRH committed:
Refactor GIsel lowering

vikramRH committed:
fix documentation mess

arsenm commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -1170,6 +1170,23 @@ The AMDGPU backend implements the following LLVM IR intrinsics.
 
+  llvm.amdgcn.readfirstlane                        Provides direct access to v_readfirstlane_b32. Returns the value in
+                                                   the lowest active lane of the input operand. Currently implemented
+                                                   for i16, i32, float, half, bf16, <2 x i16>, <2 x half>, <2 x bfloat>,
```
bf16->bfloat

vikramRH replied:
done

arsenm commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -1208,7 +1225,7 @@ The AMDGPU backend implements the following LLVM IR intrinsics.
 
   llvm.amdgcn.sdot2                                Provides direct access to v_dot2_i32_i16 across targets which
-                                                   support such instructions. This performs signed dot product
+                                                   upport such instructions. This performs signed dot product
```
Stray typo introduced

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -5387,6 +5387,98 @@ bool AMDGPULegalizerInfo::legalizeDSAtomicFPIntrinsic(LegalizerHelper &Helper,
+
+  SmallVector<Register, 2> PartialRes;
+  unsigned NumParts = Size / 32;
+  MachineInstrBuilder Src0Parts, Src2Parts;
+  Src0Parts = B.buildUnmerge(PartialResTy, Src0);
```
Fold declare + initialize

arsenm reviewed (commented):
lgtm with a few nits

vikramRH committed:
review comments

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -5461,8 +5461,7 @@ bool AMDGPULegalizerInfo::legalizeLaneOp(LegalizerHelper &Helper,
   unsigned NumParts = Size / 32;
-  MachineInstrBuilder Src0Parts, Src2Parts;
-  Src0Parts = B.buildUnmerge(PartialResTy, Src0);
+  MachineInstrBuilder Src0Parts = B.buildUnmerge(PartialResTy, Src0), Src2Parts;
```
Don't also declare Src2Parts on the same line

vikramRH replied:
Done

vikramRH committed:
handle comment

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -5496,6 +5496,9 @@ const char* AMDGPUTargetLowering::getTargetNodeName(unsigned Opcode) const {
   NODE_NAME_CASE(LDS)
   NODE_NAME_CASE(FPTRUNC_ROUND_UPWARD)
   NODE_NAME_CASE(FPTRUNC_ROUND_DOWNWARD)
+  NODE_NAME_CASE(READLANE)
+  NODE_NAME_CASE(READFIRSTLANE)
```
Add these to `AMDGPUTargetLowering::isSDNodeAlwaysUniform`

vikramRH replied:
done

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -5496,6 +5496,9 @@ const char* AMDGPUTargetLowering::getTargetNodeName(unsigned Opcode) const {
   NODE_NAME_CASE(FPTRUNC_ROUND_DOWNWARD)
+  NODE_NAME_CASE(READLANE)
+  NODE_NAME_CASE(READFIRSTLANE)
+  NODE_NAME_CASE(WRITELANE)
```
Add this to `SITargetLowering::isSDNodeSourceOfDivergence`

jayfoad reviewed (commented):
Does this need IR autoupgrade?

arsenm commented:
> Does this need IR autoupgrade?

This type of auto upgrade is free, it just happens

jayfoad commented:
There is a latent problem to do with convergence. If you add a new test case like this:
```diff
diff --git a/llvm/test/CodeGen/AMDGPU/convergence-tokens.ll b/llvm/test/CodeGen/AMDGPU/convergence-tokens.ll
index 238f6ab39e83..22995083293d 100644
--- a/llvm/test/CodeGen/AMDGPU/convergence-tokens.ll
+++ b/llvm/test/CodeGen/AMDGPU/convergence-tokens.ll
@@ -55,6 +55,21 @@ else:
   ret i32 %p
 }
 
+define i64 @basic_branch_i64(i64 %src, i1 %cond) #0 {
+entry:
+  %t = call token @llvm.experimental.convergence.anchor()
+  %x = add i64 %src, 1
+  br i1 %cond, label %then, label %else
+
+then:
+  %r = call i64 @llvm.amdgcn.readfirstlane.i64(i64 %x) [ "convergencectrl"(token %t) ]
+  br label %else
+
+else:
+  %p = phi i64 [%r, %then], [%x, %entry]
+  ret i64 %p
+}
+
 ; CHECK-LABEL: name:            basic_loop
 ;       CHECK:    [[TOKEN:%[0-9]+]]{{[^ ]*}} = CONVERGENCECTRL_ANCHOR
 ;       CHECK:  bb.[[#]].loop:
```
Then it will fail with:
```
*** Bad machine code: Cannot mix controlled and uncontrolled convergence in the same function. ***
```
This is related to #87509. Since the readlane/readfirstlane/writelane intrinsics are IntrConvergent, the corresponding ISD nodes should be marked with SDNPInGlue or SDNPOptInGlue. @ssahasra FYI

vikramRH committed:
Review comments

arsenm commented:
You should add the mentioned convergence-tokens.ll test function

vikramRH committed:
test for convergence related crash

vikramRH committed:
Update convergence-laneops-xfail.ll

vikramRH commented:
> You should add the mentioned convergence-tokens.ll test function

Added the test in a separate test file

vikramRH committed:
Merge branch 'main' into rw_lane_64

arsenm commented on llvm/test/Verifier/convergence-laneops-xfail.ll:
```diff
@@ -0,0 +1,19 @@
+; RUN: not --crash llc -stop-after=amdgpu-isel -mtriple=amdgcn-- -mcpu=gfx900 -verify-machineinstrs -o - %s 2>&1 | FileCheck %s
```
This is not an IR verifier test, it is a codegen test that fails the machine verifier. A machine verifier would go in test/MachineVerifier, and preferably would be written in MIR and not rely on codegen.


I thought the point of this test was to show that the selection did not produce a machine verifier error after selection, so this is broken

arsenm replied:
This should also be repeated for all 3 intrinsics

arsenm reviewed (changes requested):
@jayfoad's testcase fails and the same test should be repeated for all 3 intrinsics

vikramRH committed:
update convergence related failure tests

arsenm commented on llvm/test/MachineVerifier/convergencectrl/AMDGPU/convergence-readfirstlane64-xfail.mir:
```diff
@@ -0,0 +1,46 @@
+# RUN: not --crash llc -mtriple=amdgcn -run-pass=none -verify-machineinstrs -o /dev/null %s 2>&1 | FileCheck %s
```
You should not need to introduce any new machine verifier tests, they are not useful. The useful test would be the IR that produces the verifier error, which ideally would be fixed

vikramRH replied:
Okay, I'll update with IR's

ssahasra replied:
All it needs is one new file in `test/CodeGen/AMDGPU` where 64-bit lane ops are used with a convergence tokens. Mark that as XFAIL. When the issue is fixed, that file can be merged into the existing tests. We don't need to test each of the convergence control intrinsics. It's enough to just have a token on a readlane.

arsenm replied:
I'd still test all 3, but yes an IR test

ssahasra commented:
> > @jayfoad's testcase fails and the same test should be repeated for all 3 intrinsics
> 
> added MIR tests for 3 intrinsics. The issue is that Im not able to attach the glue nodes to newly created laneop pieces since they fail at selection. #87509 should enable this,

I am not really comfortable waiting for #87509 to fix convergence tokens in this expansion. Is it really true that this expansion cannot be fixed independent of future work on `CONVERGENCE_GLUE`? There is no way to manually handle the same glue operands??

vikramRH commented:
> > > @jayfoad's testcase fails and the same test should be repeated for all 3 intrinsics
> > 
> > 
> > added MIR tests for 3 intrinsics. The issue is that Im not able to attach the glue nodes to newly created laneop pieces since they fail at selection. #87509 should enable this,
> 
> I am not really comfortable waiting for #87509 to fix convergence tokens in this expansion. Is it really true that this expansion cannot be fixed independent of future work on `CONVERGENCE_GLUE`? There is no way to manually handle the same glue operands??

I guess one way would be to have custom selection for each of the new node type introduced, but would this be a proper way forward ? (this would be in general for all convergent SDNodes i guess if selection is not made generic)

vikramRH commented:
> > > > @jayfoad's testcase fails and the same test should be repeated for all 3 intrinsics
> > > 
> > > 
> > > added MIR tests for 3 intrinsics. The issue is that Im not able to attach the glue nodes to newly created laneop pieces since they fail at selection. #87509 should enable this,
> > 
> > 
> > I am not really comfortable waiting for #87509 to fix convergence tokens in this expansion. Is it really true that this expansion cannot be fixed independent of future work on `CONVERGENCE_GLUE`? There is no way to manually handle the same glue operands??
> 
> I guess one way would be to have custom selection for each of the new node type introduced, but would this be a proper way forward ? (this would be in general for all convergent SDNodes i guess if selection is not made generic)

Or drop the new nodes altogether and legelaize to intrinsics directly ?

vikramRH committed:
Merge branch 'main' into rw_lane_64

vikramRH committed:
revert targte specific SDNodes, handle convergence tokens in SDAG

vikramRH committed:
remove spurious new lines

vikramRH commented:
> That's another option. The only real plus to the intermediate is it's slightly less annoying to write combines for. But there are limited combining opportunities for these

we now legalize to intrinsics directly. The SDAG lowering uses a new helper to unroll vector cases while also handling convergence tokens

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -6129,13 +6150,55 @@ static SDValue lowerLaneOp(const SITargetLowering &TLI, SDNode *N,
+    SmallVector<SDValue, 4> Operands(NumOperands);
+    SDNode *GL = N->getGluedNode();
+
+    if (GL) {
+      // only handle convegrencectrl_glue
+      assert(GL->getOpcode() == ISD::CONVERGENCECTRL_GLUE);
+    }
```
Typo convegrencectrl_glue. Also assert(!GL || ...)

arsenm commented on llvm/test/CodeGen/AMDGPU/convergence-laneops.ll:
```diff
@@ -0,0 +1,65 @@
+; RUN: llc -stop-after=amdgpu-isel -mtriple=amdgcn-- -mcpu=gfx1100 -verify-machineinstrs -o - %s | FileCheck --check-prefixes=CHECK,ISEL %s
+
+; CHECK-LABEL: name:            basic_readfirstlane_i64
+;       CHECK:    [[TOKEN:%[0-9]+]]{{[^ ]*}} = CONVERGENCECTRL_ANCHOR
```
We don't need these isel checks for the tokens if the IR verifier is going to catch the breakage. This test can jus be merged with the existing intrinsic tests

vikramRH replied:
I currently see machine verifier failure which is not related to this patch. An  i32 example with trunc here, https://godbolt.org/z/he8asMe77. 
This is also seen with wider type legalizations that we do now, so I cannot integrate these with existing tests just yet. am I missing something here ?

arsenm replied:
I thought you switched to using the intrinsics such that this would avoid the verifier error. This test is going to fail if you are still hitting the error

vikramRH replied:
this is a preexisting error, and the failure is further down the pipeline. (after sreg alloc now i guess), does it make sense to have it as xfail now rather then stopping after isel?

arsenm replied:
My worry is this test won't get updated whenever that's fixed. How about you leave the existing checks, but add an additional "not --crash" run line that checks the verifier error?

vikramRH replied:
Makes sense, updated.

vikramRH committed:
review comments

vikramRH committed:
Merge branch 'main' into rw_lane_64

vikramRH committed:
update builtin CodeGen
